### ISSUE TITLE
`OptimisticResetVecEnvWrapper` silently corrupts states when `sum(done) < num_resets`

### ISSUE THREAD

**Issue by Pabloo22** [MichaelTMatthews/Craftax_Baselines#11]
In `OptimisticResetVecEnvWrapper`, the `step` function uses `jax.random.choice` to select which environments receive the limited pool of optimistic resets. However, when the number of environments that actually terminated (`sum(done)`) is less than `self.num_resets` (which happens frequently, especially at the start of training), the wrapper silently fails, corrupting the reset distribution.

The issue stems from this block in the `step` function:

```python
being_reset = jax.random.choice(
    _rng,
    jnp.arange(self.num_envs),
    shape=(self.num_resets,),
    p=done, 
    replace=False,
)
```

When `replace=False`, `jax.random.choice` relies on a Gumbel-max trick. If the `done` array has fewer non-zero entries than the requested `shape` (i.e., fewer environments terminated than `num_resets`), JAX cannot fulfill the `replace=False` constraint. As noted in the JAX documentation, the output of this function becomes mathematically ill-defined. Instead of raising an error during JIT compilation or runtime, it silently produces garbage indices, assigning duplicated or entirely incorrect reset states to environments.

### Proposed Solution
The new logic works by:
1. Assigning random uniform priorities to `done` environments, and `-inf` to environments that are not done.
2. Using `jax.lax.top_k` to extract the top `num_resets` candidates.
3. Creating a boolean mask (`> -jnp.inf`) to ensure we only apply resets to environments that actually terminated.
4. Using `jax.lax.fori_loop` to safely and cleanly map the generated reset states into the fallback `reset_indexes` array.

I have this fix tested locally. I am happy to open a PR with this solution if the maintainers agree with the approach!

**Comment by MichaelTMatthews** [comment]
Thanks for raising this - I was not aware of this behaviour from jax.random.choice.

From what I can tell with the current JAX implementation, in practice the invalid/arbitrary indices would all be for done=False workers, for which the env states get discarded anyway.

But you are correct to point out that this is making use of technically undefined behaviour and could break in future jax versions. I have added a simple fix for this here: https://github.com/MichaelTMatthews/Craftax_Baselines/commit/7ce36fa05b84a2c9e758012f1e6da402e1e3a891
